src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs

```diff
@@ -0,0 +1,191 @@
+            }
+
+            for (var i = 0; i < model.Filters.Count; i++)
+            {
+                filters.Add(new FilterDescriptor(model.Filters[i], FilterScope.Action));
+            }
+
+            foreach (var selector in model.Selectors)
+            {
+                actions.Add(new PageActionDescriptor()
+                {
+                    AttributeRouteInfo = new AttributeRouteInfo()
+                    {
+                        Name = selector.AttributeRouteModel.Name,
+                        Order = selector.AttributeRouteModel.Order ?? 0,
+                        Template = selector.AttributeRouteModel.Template,
+                    },
+                    DisplayName = $"Page: {item.Path}",
+                    FilterDescriptors = filters,
+                    Properties = new Dictionary<object, object>(model.Properties),
+                    RelativePath = item.CombinedPath,
+                    RouteValues = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
+                    {
+                        { "page", item.PathWithoutExtension },
+                    },
+                    ViewEnginePath = item.Path,
+                });
+            }
+        }
+
+        private RouteTemplates GetRouteTemplates(RazorProjectItem item)
+        {
+            string template;
+            if (!PageDirectiveFeature.TryGetRouteTemplate(item, out template))
+            {
+                return default(RouteTemplates);
+            }
+
+            Debug.Assert(template != null);
+
+            if (template.Length > 0 && template[0] == '/')
+            {
+                template = template.Substring(1);
+            }
+            else if (template.Length > 1 && template[0] == '~' && template[1] == '/')
+            {
+                template = template.Substring(2);
+            }
+
+            var filePath = item.CombinedPathWithoutExtension.Substring(1);
+            var routeTemplate = GetRouteTemplate(filePath, template);
+
+            if (string.Equals("Index.cshtml", item.Filename, StringComparison.OrdinalIgnoreCase))
+            {
+                return new RouteTemplates(routeTemplate, GetRouteTemplate(item.BasePath.Substring(1), template));
+            }
+
+            return new RouteTemplates(routeTemplate);
+        }
+
+        private static string GetRouteTemplate(string prefix, string template)
+        {
+            if (prefix == string.Empty && string.IsNullOrEmpty(template))
+            {
+                return string.Empty;
+            }
+            else if (string.IsNullOrEmpty(template))
+            {
+                return prefix;
+            }
+            else if (prefix == string.Empty)
+            {
+                return template;
+            }
+            else
+            {
+                return prefix + "/" + template;
+            }
+        }
+
+        private struct RouteTemplates
```

Comment:
This is kinda gross, and isn't close to what's really being done here. My suggestion would be to pass in the `PageModel` and add `SelectorModels` to it.